scene/3d/path.cpp

```diff
@@ -329,3 +339,108 @@ PathFollow::PathFollow() {
 	cubic = true;
 	loop = true;
 }
+
+//////////////
+
+OrientedPathFollow::OrientedPathFollow() :
+		PathFollow() {
+}
+
+void OrientedPathFollow::_update_transform() {
+
+	// Suggestion, make some PathFollow properties protected
+
+	Path *path = _get_path();
+	if (!path)
+		return;
+
+	Ref<Curve3D> c = path->get_curve();
+	if (!c.is_valid())
+		return;
+
+	int count = c->get_point_count();
+	if (count < 2)
+		return;
+
+	if (_get_delta_offset() == 0) {
+		return;
+	}
+
+	float offset = get_offset();
+	float bl = c->get_baked_length();
+	float bi = c->get_bake_interval();
+	float o = offset;
+	float o_next = offset + bi;
+
+	if (has_loop()) {
+		o = Math::fposmod(o, bl);
+		o_next = Math::fposmod(o_next, bl);
+	} else if (o_next >= bl) {
+		o = bl - bi;
+		o_next = bl;
+	}
+
+	bool cubic = get_cubic_interpolation();
+	Vector3 pos = c->interpolate_baked(o, cubic);
+	Vector3 forward = c->interpolate_baked(o_next, cubic) - pos;
+
+	if (forward.length_squared() < CMP_EPSILON2)
+		forward = Vector3(0, 0, 1);
+	else
+		forward.normalize();
+
+	RotationMode rotation_mode = get_rotation_mode();
+	bool apply_tilt = rotation_mode == ROTATION_XYZ;
+	Vector3 up = c->interpolate_baked_up_vector(o, apply_tilt);
+
+	if (o_next < o) {
+		Vector3 up1 = c->interpolate_baked_up_vector(o_next, apply_tilt);
+		Vector3 axis = up.cross(up1);
+
+		if (axis.length_squared() < CMP_EPSILON2)
+			axis = forward;
+		else
+			axis.normalize();
+
+		up.rotate(axis, up.angle_to(up1) * 0.5f);
+	}
+
+	Transform t = get_transform();
+
+	Vector3 sideways = up.cross(forward).normalized();
+	up = forward.cross(sideways).normalized();
+
+	if (rotation_mode != ROTATION_NONE) {
+
+		Vector3 scale = t.basis.get_scale();
+
+		Basis basis;
+		basis.set(sideways, up, forward);
+
+		// I will not use set_euler_scale directly here because I need to update sideways and up with the new basis info. (in case a constraint is set.)
+		if (rotation_mode == ROTATION_Y) {
+			Vector3 euler = basis.get_euler();
+			euler.x = 0.0f;
+			euler.z = 0.0f;
+			basis.set_euler(euler);
+
+			sideways = basis.get_axis(0);
+			up = basis.get_axis(1);
+		} else if (rotation_mode == ROTATION_XY) {
+			Vector3 euler = basis.get_euler();
+			euler.z = 0.0f;
+			basis.set_euler(euler);
+
+			sideways = basis.get_axis(0);
+			up = basis.get_axis(1);
+		}
+
+		// does exactly what set_scale does, but without unnecessary normalizing.
```

Comment:
mmm I guess we're talking about different methods, I was referring to set_scale whose definition is:

`
void Basis::set_scale(const Vector3 &p_scale) {

	set_axis(0, get_axis(0).normalized() * p_scale.x);
	set_axis(1, get_axis(1).normalized() * p_scale.y);
	set_axis(2, get_axis(2).normalized() * p_scale.z);
}
`

but I see scale works fine as well, so I'll replace with it.

Reply:
Ah, thanks for pointing this out, that set_scale is wrong (it's setting scale in object-local frame, whereas it should be doing that in the parent-local, or "global" within Basis class), and I'll submit a PR to remove it (I removed a correct version of it in the past because it's just confusing for people who use negative scalings). I'll also add the missing Basis constructor.

Reply:
Hold on, sorry, scaling should happen before rotation, so what set_scale did was right when scales are positive (although still wrong with negative scales). You need to use scale_local to achieve the same effect

Reply:
scale seems to work fine, could you give a code sample of how scale_local would achieve the same effect?

Reply:
done.